File: Js_Solutions/hash_table/hash_table.js
// Implement a hash table

class HashTable {
    constructor(size = 53) {
        this.keyMap = new Array(size);
    }

    _hash(key) {
        let total = 0;
        const WEIRD_PRIME = 31;
        for (let i = 0; i < Math.min(key.length, 100); i++) {
            let char = key[i];
            let value = char.charCodeAt(0) - 96;
            total = (total * WEIRD_PRIME + value) % this.keyMap.length;
        }
        return total;
    }

    set(key, val) {
        let idx = this._hash(key);
        if (!this.keyMap[idx]) {
            this.keyMap[idx] = [];
        }
        this.keyMap[idx].push([key, val])
        return this.keyMap;
    }

    get(key) {
        let idx = this._hash(key);
        if (this.keyMap[idx]) {
            for (let i = 0; i < this.keyMap[idx].length; i++) {
                if (this.keyMap[idx][i][0] === key) {
                    return this.keyMap[idx][i][1];
                }
            }
        }
        return undefined;
    }

    keys() {
        let result = [];
        for (let i = 0; i < this.keyMap.length; i++) {
            if (this.keyMap[i]) {
                for (let j = 0; j < this.keyMap[i].length; j++) {
                    result.push(this.keyMap[i][j][0])
                }
            }
        }
        return result;
    }

    values() {
        let result = [];
        for (let i = 0; i < this.keyMap.length; i++) {
            if (this.keyMap[i]) {
                for (let j = 0; j < this.keyMap[i].length; j++) {
                    if (!result.includes(this.keyMap[i][j][1])) {
                        result.push(this.keyMap[i][j][1])
                    }
                }
            }
        }
        return result;
    }

    items() {
        let result = [];
        for (let i = 0; i < this.keyMap.length; i++) {
            if (this.keyMap[i]) {
                for (let j = 0; j < this.keyMap[i].length; j++) {
                    result.push(this.keyMap[i][j]);
                }
            }
        }
        return result;
    }
}

h = new HashTable();
h.set('pink', 123);
console.log(h.set('cyan', 231));
console.log(h.set('tomato', 111));
console.log(h.set('orange', 1));
console.log(h.set('hello world', 2));
console.log(h.get('orange'));
console.log(h.keys());
console.log(h.values());
console.log(h.items());

File: Python_Solutions/hash_table/hash_table.py
# Implement a hash table

class HashTable:
    def __init__(self, size=53):
        self.keyMap = [None for i in range(size)]

    def _hash(self, key):
        total = 0
        FIXED_PRIME = 31
        for i in range(min(len(key), 100)):
            value = ord(key[i]) - 96
            total = (total * FIXED_PRIME + value) % len(self.keyMap)
        return total

    def set_val(self, key, val):
        idx = self._hash(key)
        if self.keyMap[idx] is None:
            self.keyMap[idx] = []
        self.keyMap[idx].append([key, val])
        return True

    def get(self, key):
        idx = self._hash(key)
        if self.keyMap[idx] is not None:
            for i in range(len(self.keyMap[idx])):
                if self.keyMap[idx][i][0] == key:
                    return self.keyMap[idx][i][1]
        return None

    def keys(self):
        result = []
        for i in range(len(self.keyMap)):
            if self.keyMap[i] is not None:
                for j in range(len(self.keyMap[i])):
                    result.append(self.keyMap[i][j][0])
        return result

    def values(self):
        result = []
        for i in range(len(self.keyMap)):
            if self.keyMap[i] is not None:
                for j in range(len(self.keyMap[i])):
                    result.append(self.keyMap[i][j][1])
        return result

    def items(self):
        result = []
        for i in range(len(self.keyMap)):
            if self.keyMap[i] is not None:
                for j in range(len(self.keyMap[i])):
                    result.append(self.keyMap[i][j])
        return result


h = HashTable()
h.set_val('pink', 123)
print(h.set_val('cyan', 231))
print(h.set_val('tomato', 111))
print(h.set_val('orange', 1))
print(h.set_val('hello world', 2))
print(h.get('orange'))
print(h.keys())
print(h.values())
print(h.items())
